### PR TITLE
Support ELF format payload

### PR DESCRIPTION
**Pull request by myguitar**
In order to support ELF format payload in Stage2, ElfLib is
separated from MultibootLib.

Change-Id: Iac45e8511101e2d12cb9a9f7f9bb2394a1e7be39
Signed-off-by: Aiden Park <aiden.park@intel.com>

### CODE REVIEW

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+  Eaddr = 0;
+  for (Index = 0; Index < Ehdr->E_Phnum ; Index ++) {
+    if (Phdr[Index].P_Type != PT_LOAD) {
+      continue;
+    }
+    Eaddr = MAX (Eaddr, Phdr[Index].P_Paddr + Phdr[Index].P_Memsz);
+  }
+
+  MbInfo->Flags       |= MULTIBOOT_INFO_HAS_ELF_SYMS;
```
Comment (mborgerson) [inline BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: I think the `MULTIBOOT_INFO_HAS_ELF_SYMS` name is a bit misleading and does not exactly line up with [the spec](https://www.gnu.org/software/grub/manual/multiboot/multiboot.html) which defines this bit as:
```
     /* is there an ELF section header table? */
     #define MULTIBOOT_INFO_ELF_SHDR                 0X00000020
```

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  Shdr = (ELF32_SHDR *) (UINTN) Eaddr;
+  Size = Ehdr->E_Shnum * Ehdr->E_Shentsize;
+  CopyMem (Shdr, (VOID *) ((UINT32)ImageBase + Ehdr->E_Shoff), Size);
+  Eaddr += Size;
+
+  // Copy sections that are part of the symbol table.
+  for (Index = 0 ; Index < Ehdr->E_Shnum ; Index += 1) {
+    if (Shdr[Index].Sh_Type != SHT_SYMTAB) {
```
Comment (mborgerson) [inline BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: Two questions here: If we are loading the section header table...why not load the entire thing instead of just the symbol table? If we are loading just the symbol table..why should we load it at all? Does any kernel actually depend on the symbol table being loaded?

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.inf [hunk 3]
```diff
@@ -0,0 +1,43 @@
+## @file
+#
+#  Copyright (c) 2017, Intel Corporation. All rights reserved.<BR>
```
Comment (mborgerson) [inline BootloaderCommonPkg/Library/ElfLib/ElfLib.inf]: Year

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  ELF32_PHDR                 *Phdr;
+  UINT32                      Eaddr;
+  UINT32                      Index;
+  UINT32                      Msize;
+  UINT32                      Fsize;
+  UINT32                      Mpos;
+  UINT32                      Ipos;
+
+  //
+  // TODO: Check if image, load region overlap; if yes, move image away.
```
Comment (mborgerson) [inline BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: Perhaps this can be deferred to another PR, but we should really do some bounds checking to make sure we can actually load the ELF segments into memory at the specified locations. Do we have enough memory? Have the ranges been reserved for something else? Etc.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 1 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: Yes, it seems. Let me include the change in next commit rather than new PR.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 2 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: All necessary things for execution binary are loaded thru program header. Only symbol information needs to be loaded thru section header. If we do not use program header, we can load the entire thing thru section header.
- Code from program header
- Symbol from section header
This is a very minimal way to load executable code and symbol because program header is much simpler than section header.

I am not sure kernel depends on the symbol table, but it may be optional for debugging. If we don't want to load the information, we can strip ELF.

Someone has any idea?

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.inf (hunk 3 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.inf]: Let me change it.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 4 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: Yes, agree and let's defer this. PE format image relocation as well.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 1 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: I double checked it. It may need another PR since other bits need to be updated as well. Let me use the name as it is in this commit.

File: BootloaderCommonPkg/Include/Library/ElfLib.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#ifndef __ELF_LIB_H__
+#define __ELF_LIB_H__
+
+#include <Uefi/UefiBaseType.h>
+#include <Library/MultibootLib.h>
```
Comment (mauricema) [inline BootloaderCommonPkg/Include/Library/ElfLib.h]: can we decouple Multiboot and ELF?   Why ELF needs to include MultiBoot?   Shouldn't be the other way ?

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 2 above)
Comment (mauricema) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: dynamic symbols could potentially be used at RT for elf.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+  @param[in]        ImageBase   Memory address of an image
+  @param[in, out]   MbInfo      Pointer to MULTIBOOT_INFO which needs to be updated
+
+  @retval RETURN_SUCCESS        MULTIBOOT_INFO successfully updated.
+  @retval RETURN_NOT_FOUND      No ELF Symbol information found.
+
+**/
+RETURN_STATUS
+UpdateElfMultibootInfo (
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: I think we should try to move MULTBOOT related code into Multiboot library.  Here if required, create new function to return ELF info to be populated into MULTIBOOT_INFO.

Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: Okay, let me create a new function for ELF info and use it in MultibootLib.

File: BootloaderCommonPkg/Include/Library/ElfLib.h (hunk 5 above)
Comment (myguitar) [reply BootloaderCommonPkg/Include/Library/ElfLib.h]: I will do that.

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 2 above)
Comment (myguitar) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: SHT_DYNSYM(dynamic symbols) is a smaller version of SHT_SYMTAB(symbols). SYMTAB already includes DYNSYM.

File: PayloadPkg/OsLoader/OsLoader.c [hunk 7]
```diff
@@ -208,11 +208,12 @@ SetupBootImage (
   MultiBoot = &LoadedImage->Image.MultiBoot;
   if (IsElfImage (BootFile->Addr)) {
     DEBUG ((DEBUG_INFO, "Boot image is ELF format...\n"));
-    EntryPoint = LoadElfImage (MultiBoot);
+    EntryPoint = LoadElfImage (BootFile->Addr);
     if (EntryPoint != NULL) {
       if (IsMultiboot (BootFile->Addr)) {
         DEBUG ((DEBUG_INFO, "and Image is Multiboot format\n"));
         SetupMultibootInfo (MultiBoot);
+        UpdateElfMultibootInfo (BootFile->Addr, &MultiBoot->MbInfo);
```
Comment (gdong1) [inline PayloadPkg/OsLoader/OsLoader.c]: Prefer add UpdateElfMultibootInfo() into SetupMultibootInfo() to make this file clean. both parameters in new function could be got from MultiBoot parameter of SetupMultibootInfo().

File: BootloaderCommonPkg/Library/ElfLib/ElfLib.c (hunk 2 above)
Comment (mborgerson) [reply BootloaderCommonPkg/Library/ElfLib/ElfLib.c]: What I mean here is that we are providing ELF section information through the multiboot interface, but only the symbol section. It could be the case that the OS depends on other section headers as well (.text/.data/etc). Should we not also load these if we are doing this?

File: PayloadPkg/OsLoader/OsLoader.c (hunk 7 above)
Comment (gdong1) [reply PayloadPkg/OsLoader/OsLoader.c]: I am OK with that. For all multiple boot support to ELF image, we put it in SetupMultibootInfo(). we could enhance this function later if we really need it.

**Comment by myguitar** [comment]
Windows build broken - one more patch is coming.
